### ISSUE TITLE
all_smiles_XX files in morgan_1024_predictions folder

### ISSUE THREAD

**Issue by albertma-evotec** [jamesgleave/Deep-Docking-NonAutomated#13]
Hi

I have completed a DD run of 8 iterations. I found that in the morgan_1024_predictions folder in each iteration folder, the total number of compound predicted is not always the same. I thought it always predict the whole dataset at phase 3. The number of compound to be predict should always be the same as the size of the DB, isnt that true?
Here are I get the sum of all compounds in the smiles_all_XX files for each iteration and for the database as well.
![image](https://user-images.githubusercontent.com/54853041/123283224-1a6e8180-d503-11eb-99ae-7e7dd594ec16.png)

My database size is 3.798M

Thanks

**Comment by fgentile89** [comment]
Hi,
no the number always change (and should decrease). The reason is that, while inference is always run on the full database, the morgan_1024_predictions folder just has compounds that are positively classified (probability above the threshold value in thresholds.txt which should ensure 90% recall or any recall you specified), while the non-hits are discarded (in this way, your database becomes smaller and smaller while retaining 90% of the top scoring molecules). As the model improves and the condition to be a virtual hit keeps becoming more stringent, the program is able to discard more unqualified compounds. You can actually see that the in the files, the second column has a float value. This is the probability from inference, and they are all above the cutoff one.

I see that between iteration 7 and 8 the number of molecules increases. It could be because you use a smaller sampling size and there are not enough true hits in validation/test/training sets to guarantee a generalizable model, perhaps? I would say you have only 30 positives at that point (check the last value of each line in hyperparameter...csv file)?

Cheers

Francesco

**Comment by albertma-evotec** [comment]
Thanks for the prompty reply and the explanation. 

As I do not have enough to resource to run a retrospective study on a ultra large DB, I am using a much smaller one (~3.8M). It could be the reason. In fact, knowing my db is considerably smaller, I am using percent_first_mols=0.05 and percent_last_mol=0.005 instead of the reported 0.01 and 0.0001 in the paper to try to get enough hit. My sampling size is 9999 for each train/valid/test set in iteration1 and I intended to sample 10000 in total for subsequent steps (phase 1) but its actual size is always below that and it fluctuated.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ama/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ama/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | Train size | Valid size | Test size
-- | -- | -- | --
1 | 9999 | 9999 | 9999
2 | 3271 | 3274 | 3265
3 | 3243 | 3247 | 3262
4 | 3249 | 3243 | 3238
5 | 3221 | 3215 | 3231
6 | 3211 | 3195 | 3203
7 | 3203 | 3200 | 3168
8 | 3176 | 3154 | 3184



</body>

</html>

This is the hyperparameter_morgan_with_freq_v3.csv file of the last iteration:
![image](https://user-images.githubusercontent.com/54853041/123300483-8a383880-d512-11eb-8231-586a8b98008f.png)

I believe 56 is the number of true hit you meant.

**Comment by fgentile89** [comment]
Yes exactly, 56 is a little bit too low and the models suffer from that (you can also see that the recall in test set, second last column, is quite variable and not always close to the expected value of 0.9. Perhaps (but I've never tried my self as I deal with larger datasets), you can modify line 90 of simple_job_models.py (oversampling) and set it to 20, to see if it has any effect on generalizability.

Francesco
